zebra/zebra_ted.c

```diff
@@ -0,0 +1,280 @@
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+
+#include "lib/zebra.h"
+#include "lib/debug.h"
+#include "lib/frr_pthread.h"
+#include "lib/stream.h"
+#include "zebra/debug.h"
+#include "zebra/zserv.h"
+#include "zebra/zebra_ted.h"
+
+/*
+ * Globals
+ */
+static struct zebra_ted_globals {
+
+	/* Sentinel for beginning of shutdown */
+	volatile bool is_shutdown;
+
+	/* Sentinel for run or start of shutdown */
+	volatile bool run;
+
+	/* Limit number of pending, unprocessed updates */
+	_Atomic uint32_t max_queued_updates;
+
+	/* Limit number of new messages dequeued at once, to pace an
+	 * incoming burst.
+	 */
+	uint32_t msgs_per_cycle;
+
+	/* Stats: counters of incoming messages, errors, and yields (when
+	 * the limit has been reached.)
+	 */
+	_Atomic uint32_t msgs_in;
+	_Atomic uint32_t msg_errors;
+	_Atomic uint32_t yields;
+
+	/* pthread */
+	struct frr_pthread *pthread;
+
+	/* Event-delivery context 'master' for the module */
+	struct thread_master *master;
+
+	/* Event/'thread' pointer for queued zapi messages */
+	struct thread *t_msgs;
+
+	/* Input fifo queue to the module, and lock to protect it. */
+	pthread_mutex_t mutex;
+	struct stream_fifo in_fifo;
+
+} zt_info;
+
+/* Name string for debugs/logs */
+static const char *LOG_NAME = "Zebra TED";
+
+/* Prototypes */
+
+/* Main event loop, processing incoming message queue */
+static int process_messages(struct thread *event);
+
+/*
+ * Initialize the module at startup
+ */
+void zebra_ted_init(void)
+{
+	memset(&zt_info, 0, sizeof(zt_info));
+
+	pthread_mutex_init(&zt_info.mutex, NULL);
+	stream_fifo_init(&zt_info.in_fifo);
+
+	zt_info.msgs_per_cycle = ZEBRA_TED_MSG_LIMIT;
+}
+
+/*
+ * Start the module pthread. This step is run later than the
+ * 'init' step, in case zebra has fork-ed.
+ */
+void zebra_ted_start(void)
+{
+	struct frr_pthread_attr pattr = {
+		.start = frr_pthread_attr_default.start,
+		.stop = frr_pthread_attr_default.stop
+	};
+
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s module starting", LOG_NAME);
+
+	/* Start pthread */
+	zt_info.pthread = frr_pthread_new(&pattr, "Zebra TED thread",
+					  "zebra_TED");
+
+	/* Associate event 'master' */
+	zt_info.master = zt_info.pthread->master;
+
+	zt_info.run = true;
+
+	/* Enqueue an initial event for the pthread */
+	thread_add_event(zt_info.master, process_messages, NULL, 0,
+			 &zt_info.t_msgs);
+
+	/* And start the pthread */
+	frr_pthread_run(zt_info.pthread, NULL);
+}
+
+/*
+ * Module stop, halting the dedicated pthread; called from the main pthread.
+ */
+void zebra_ted_stop(void)
+{
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s module stop", LOG_NAME);
+
+	zt_info.run = false;
+
+	frr_pthread_stop(zt_info.pthread, NULL);
+
+	THREAD_OFF(zt_info.t_msgs);
```

Comment:
Not only is this unnecessary - you already killed the whole pthread and so the event loop isn't even running - it's also not thread safe even if it were running, you would need to use an async cancel.